### DISCUSSION TITLE
Hide a menu button?

### DISCUSSION THREAD

**Discussion by Deanobeano2** [cortexrd/Knack-Toolkit-Library#306] [Ideas]
Hi

Is it possible to hide/display a menu button based on a User Role?

We have a menu button linking to a Bulk Operations page which we like to hide from view, or display only for Bulk Edit users.

Dean

**Comment by amandajane-mo** [comment]
Hi Dean,

Yes it is.

You can use _cls to add class ktlDisplayNone, combined with ktlRoles and
ktlTarget to target your button. Something like:

_cls=[ktlDisplayNone], [ktlRoles !Bulk Edit], [ktlTarget, $(‘#${viewId}
li:contains(“button text”))’]

Let me know how you get on.

Amanda

**Comment by Deanobeano2** [comment]
Amanda

Thank you for providing that so promptly.

That seems to hide the whole menu, of 3 buttons, for users irrespective of the Bulk Edit role.
I'm trying to hide just the 1 menu button containing Bulk Operations.
![image](https://github.com/cortexrd/Knack-Toolkit-Library/assets/130625963/dfed3ae9-ab52-4c80-aca8-79ff0f51acf8)

Dean

**Comment by amandajane-mo** [comment]
Hi Dean,

Could you send your keyword please?

Amanda

**Comment by Deanobeano2** [comment]
Hi Amanda

Never having used the keyword _cls before, I used the solution you provided and tried my view# etc and iterations of this

_cls=[ktlDisplayNone], [ktlRoles !Bulk Edit], [ktlTarget, $(‘#${#view_1109}
li:contains(“Bulk Operations”))’]

Sadly, everything I tried, not really knowing what I was doing, ended up with the same result of hiding the entire menu rather than the 1 button I would like to hide.

Dean

**Comment by amandajane-mo** [comment]
Hi Dean,

I see the problem, I should have explained the parameters better. The
viewId in {viewId} is a variable that will pull in the id of the view your
keyword is on, so #${viewId} becomes #view_xyz. This is useful if you're
copying keywords to other views as it means you don't have to keep changing
the view ID, the KTL will do that for you.

As long as your KTL is reasonably up to date using ${viewId} should work,
if not use #view_1109 with no {}.

Try:

_cls=[ktlDisplayNone], [ktlRoles !Bulk Edit], [ktlTarget, $(‘#${viewId}
.menu-links__list-item:contains("Bulk Operations")')]

Or:

_cls=[ktlDisplayNone], [ktlRoles !Bulk Edit], [ktlTarget, $(‘#view_1109
.menu-links__list-item:contains("Bulk Operations")')]

This is what I use to hide Knack menu button, should work 🤞🏻

Amanda

**Comment by Deanobeano2** [comment]
Hi Amanda

Both options still result in the entire menu being hidden irrespective of the User Role being Bulk Edit or not.

Dean

**Comment by amandajane-mo** [comment]
Hi Dean,

Looks like there's an issue with the ktlTarget which is leading to the
whole view being hidden. Could you tell me what version of ktl you're using
and send me your latest _cls.

Could you also inspect your menu (press f12, then ctrl shift c or cmd shift
c on mac) and send me a screenshot of the html structure of your menu? Mine
looks like this:

[image: Screen Shot 2024-07-01 at 11.53.50.png]

Maybe worth trying it without the ktlRoles and see if that makes any
difference.

Thanks
Amanda

**Comment by Deanobeano2** [comment]
Hi Amanda

We're running the latest V0.27.0.

I've currently removed the _cls so that we can see the menu for day to day use and I did try removing the ktlRoles but that also seemed to make no difference.

![image](https://github.com/cortexrd/Knack-Toolkit-Library/assets/130625963/3be5e76d-9a8a-44df-8a1e-74639444b3ba)

Dean

**Comment by Deanobeano2** [comment]
Hi Amanda

Sorry, you probably wanted this information.

![image](https://github.com/cortexrd/Knack-Toolkit-Library/assets/130625963/400c72f1-128e-41bd-83d8-296dc44a493c)

Dean

**Comment by amandajane-mo** [comment]
Hi Dean,

Having further inspected the ktlTarget I found an erant inverted comma (one
of them was different from the others and therefore didn't form a pair).
This meant the ktlTarget wasn't valid and therefore the whole view was
hidden, as if there was no target at all. Try this:

_cls=[ktlDisplayNone], [ktlRoles !Bulk Edit], [ktlTarget, $('#${viewId} .menu-links__list-item:contains("Bulk Operations")')]

Or

_cls=[ktlDisplayNone], [ktlRoles !Bulk Edit], [ktlTarget, $('#view_1109 .menu-links__list-item:contains("Bulk Operations")')]

I'm pretty confident this will now work, if not we may have to have a quick
zoom.

Amanda

**Comment by Greenlight88** [comment]
Try this

_cls=[ktlHidden], [ktlRoles !Bulk Edit], [ktlTarget, $('#view_1109> div > nav > ul > li:nth-child(3)')]

**Comment by cortexrd** [comment]
Hi guys,

@Greenlight88 solution should work.

This will hide any button based on its text content:

` _cls=[ktlDisplayNone], [ktlTarget, $('#view_xxx .knMenuLink:textEquals("My ButtonText")')], [ktlRoles, !Bulk Edit]`

It's safer than using the child index because it's independent of the button position.  If you move it around, it will still work.
On the other hand, if you change the text, it won't work anymore.  It's up to you to decide.

About ktlHidden vs ktlDisplayNone:  both should work and make no difference, in this case.

Norm

**Comment by Deanobeano2** [comment]
Hi All

Thank you,  @Greenlight88 that solution worked.

Norm, thank you for explaining the differences between the 2 solutions.

Dean

**Comment by Deanobeano2** [comment]
Hi

Sorry, but in my excitement that we'd achieved the result of hiding the menu button I realized that it isn't taking into account the Bulk Edit role and hides the button for all users.

I've tried [ktlRoles !Bulk Edit] and [ktlRoles, !Bulk Edit] and moving the position of this parameter but it makes no difference.

Does anyone have any thoughts on why this might be happening?

I've double-checked 2 other users accounts and when the button is present for them and they click through to the bulk operations page, which we don't want them to be able to do, they can't perform any bulk operations.

Dean

**Reply by cortexrd** [reply]
@Deanobeano2,

This is very strange... I've tested thoroughly here, and all works as expected.  Are you sure you don't have a typo in the user role or kw?  Must be case sensitive, BTW.

I used that keyword string here, and it's perfect:  
`_cls=[ktlDisplayNone], [ktlTarget, $('#view_xxx .knMenuLink:textEquals("My ButtonText")')], [ktlRoles, !Bulk Edit]`

Can you send me your web app link at this email please?  nd@ctrnd.com
No login is required.  Just the link to the App (not the Builder).

Norm

**Comment by cortexrd** [comment]
Or better if you can: create a dummy test page for me to test, with a few records and the button and the keyword.
I need no Bulk Edit role to see the bug.

**Comment by Deanobeano2** [comment]
Solved by Norm.

Make sure to follow "//Your App-specific code goes here..."
